### PR TITLE
FOUR-6098: Implement debounce for Screen text inputs after user stops typing

### PR DESCRIPTION
**Pull request by danloa**
## Note for Cypress tests:
With this PR alone, some tests will fail becaue this PR needs VFE PR (https://github.com/ProcessMaker/vue-form-elements/pull/342) merged and published so we can point to this VFE version.

## Issue & Reproduction Steps

When complex screens are created there are a number of responsiveness issues. As examples, 2 screens are attached. 
Import the 2 screens and verify that in a task:
- With the "Comprehensive" screen, typing on any text area or input line freezes the screen for several seconds.  

- The RAOS screen (the request must have the attached json as its data) send hundreds for API calls that hangs the browser and some times the nginx server.

## Solution
- Debounce added for text areas, input lines and masked inputs
- Caching of vocabularies calls 
- Caching of datasource calls

## Related Tickets & Packages
- [https://processmaker.atlassian.net/browse/FOUR-6098](https://processmaker.atlassian.net/browse/FOUR-6098)
- [https://processmaker.atlassian.net/browse/FOUR-6135](https://processmaker.atlassian.net/browse/FOUR-6135)

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

### CODE REVIEW

File: src/components/screen-renderer.vue [hunk 1]
```diff
@@ -53,6 +75,11 @@ export default {
     },
   },
   methods: {
+    hash(s) {
+      for (var i=0,h=9;i<s.length;)
+        h=Math.imul(h^s.charCodeAt(i++),9**9);
+      return h^h>>>9;
```
Comment (josechirivella) [inline src/components/screen-renderer.vue:81]: this is invalid javascript code, line 81, the h is not being picked as you think from the var declared in line 79
`h` is in the context of the for loop, not in the context of the function itself

**Review by josechirivella — CHANGES_REQUESTED** [review]

**Review by agustinbusso — COMMENTED** [review]
After some tests:

- In general, complex screens, after complete load, seems to be more responsive when typing and rendering the information.

https://user-images.githubusercontent.com/90727999/181624280-e5e182e0-2771-47da-8d96-d086155016ab.mov

- If you add some Data input in a complex screen, then replace it with {}, and paste the data again the screen takes a long time:

https://user-images.githubusercontent.com/90727999/181624583-f1a73e31-810b-440a-829d-fcd777ce7214.mov

- When you are in a complex screen and move from design to preview add Data input and then move back to design again, the screen is getting freezed, sometimes the screen get breaks: 

https://user-images.githubusercontent.com/90727999/181625023-3dfe5c1b-f437-43d2-a698-20cdb462e778.mov

Screen Shot 2022-07-28 at 16 22 13

- Extending the delay for validations to 500 ms, probably, in some scenarios can cause wrong data because you can click the submit button without validate it (is rare but I think can happen):

https://user-images.githubusercontent.com/90727999/181625474-eda9472e-e98f-4c70-8941-d308f47ee621.mov

- For cases like Comprehensive screen, after the screen are rendered, you can notice a good performance improvement when typing form inputs and text areas:

https://user-images.githubusercontent.com/90727999/181625878-20b11ce6-9bdc-4f87-90c5-a94db63648f6.mov

File: src/components/screen-renderer.vue [hunk 2]
```diff
@@ -42,7 +42,29 @@ export default {
   },
   mounted() {
     this.currentDefinition = cloneDeep(this.definition);
-    this.component = this.buildComponent(this.currentDefinition);
+
+    let screenName =  this.currentDefinition.config[0] ? this.currentDefinition.config[0].name : null;
```
[the remark this answers is not in the document]
Comment (danloa) [reply src/components/screen-renderer.vue:46]: Modified (now we use lodash get function)

File: src/components/screen-renderer.vue [hunk 3]
```diff
@@ -42,7 +42,29 @@ export default {
...
+
+    let itemName = 'hash' + this.hash(JSON.stringify(this.currentDefinition));
```
[the remark this answers is not in the document]
Comment (danloa) [reply src/components/screen-renderer.vue:48]: Done

File: src/components/screen-renderer.vue [hunk 4]
```diff
@@ -42,7 +42,29 @@ export default {
...
+
+    if (['emtpy', undefined, null].includes(screenName)) {
```
[the remark this answers is not in the document]
Comment (danloa) [reply src/components/screen-renderer.vue:50]: Done

File: src/components/screen-renderer.vue (hunk 1 above)
Comment (danloa) [reply src/components/screen-renderer.vue:81]: As in C i, h variables are outside the for block scope. At line 81, i has the string length, for example

**Comment by agustinbusso** [comment]
- Screen is not getting freeze anymore when changing from design to preview

https://user-images.githubusercontent.com/90727999/184005247-fc5cd4fa-083c-4fa2-b40d-4f5b1da232ff.mov

**Comment by agustinbusso** [comment]
With the last change I can notice a good improve in validations and submit button:

https://user-images.githubusercontent.com/90727999/184218063-ba9d4789-edd7-4909-a202-df1982294646.mov

**Review by agustinbusso — APPROVED** [review]

**Review by ryancooley — APPROVED** [review]

**Review by josechirivella — APPROVED** [review]

**Pull request merged**